### PR TITLE
Attachment fixes

### PR DESCRIPTION
**Pull request by skirill**
- bases on two previous merge requests (bugzilla login and gitlab v4 API)
- fetch attachment when it was added to the original bug description
- separate notions of "description" and "file name" same way as it works in Bugzilla

Note: I don't use format() in markdown construction because it fails for non-ascii (localized descriptions)

### CODE REVIEW

File: bugzilla2gitlab/models.py [hunk 1]
```diff
@@ -246,13 +252,13 @@ class Attachment(object):
     '''
     The attachment model
     '''
-    def __init__(self, bugzilla_attachment_id, filename):
+    def __init__(self, bugzilla_attachment_id, file_description):
         self.id = bugzilla_attachment_id
-        self.filename = filename
+        self.file_description = file_description
```
Comment (xmunoz) [inline bugzilla2gitlab/models.py]: It's a filename. You call it a filename in the `Comment` class, and it makes sense to be consistent.

File: bugzilla2gitlab/models.py [hunk 2]
```diff
@@ -262,17 +268,29 @@ def parse_filename(cls, comment):
     def save(self):
         url = "{}/attachment.cgi?id={}".format(conf.bugzilla_base_url, self.id)
         result = _perform_request(url, "get", json=False)
+        filename = self.file_description
+        # Use real filename to store attachment but descriptive name for issue text
+        if 'Content-disposition' in result.headers:
```
Comment (xmunoz) [inline bugzilla2gitlab/models.py]: I think if/else is more clear in this situation than setting the default value followed by an if statement that likely changes it. Also, wouldn't it be an error if no filename can be parsed from the `Content-disposition`?

File: bugzilla2gitlab/models.py [hunk 3]
```diff
@@ -262,17 +268,29 @@ def parse_filename(cls, comment):
...
         url = "{}/projects/{}/uploads".format(conf.gitlab_base_url, conf.gitlab_project_id)
-        f = {"file": (self.filename, result.content)}
+        f = {"file": (filename, result.content)}
 
         attachment = _perform_request(url, "post", headers=self.headers, files=f, json=True,
                                       dry_run=conf.dry_run)
 
-        if conf.dry_run:
-            return "[attachment]({})".format(self.filename)
+        link = filename
```
Comment (xmunoz) [inline bugzilla2gitlab/models.py]: Same comment as above. This pattern doesn't seem appropriate for this situation. This renaming is also generally confusing.

File: bugzilla2gitlab/models.py [hunk 4]
```diff
@@ -118,21 +118,18 @@ def create_description(self, fields):
...
         to_delete = []
-        for i in range(0, len(fields["long_desc"])):
+        comment0 = fields["long_desc"][0]
+        if (fields["reporter"] == comment0["who"] and comment0["thetext"]):
+            ext_description += "\n## Extended Description \n"
+            ext_description += "\n\n".join(re.split("\n*", comment0["thetext"]))
+            self.update_attachments(fields["reporter"], comment0, attachments)
+            to_delete.append(0)
+
+        for i in range(1, len(fields["long_desc"])):
```
Comment (xmunoz) [inline bugzilla2gitlab/models.py]: This introduces a subtle bug for the situation that the first comment is not related to an attachment, but in fact a legitimate comment (i.e. the `if` loop condition above evaluates to `false`). Please revert to `del fields["long_desc"][0]` and start this loop from 0.

**Comment by xmunoz** [comment]
Thanks for the PR! See comments above.

File: bugzilla2gitlab/models.py [hunk 5]
```diff
@@ -246,13 +252,13 @@ class Attachment(object):
...
-    def __init__(self, bugzilla_attachment_id, filename):
+    def __init__(self, bugzilla_attachment_id, file_description):
         self.id = bugzilla_attachment_id
-        self.filename = filename
+        self.file_description = file_description
         self.headers = conf.default_headers
 
     @classmethod
-    def parse_filename(cls, comment):
+    def parse_file_description(cls, comment):
```
Comment (xmunoz) [inline bugzilla2gitlab/models.py]: I think this rename is file, because you are parsing a file description to extract the name.

File: bugzilla2gitlab/models.py [hunk 6]
```diff
@@ -262,17 +268,29 @@ def parse_filename(cls, comment):
...
+        link = filename
+        if attachment and attachment["markdown"]:
+           # [mail_route.zip](/uploads/e943e69eb2478529f2f1c7c7ea00fb46/mail_route.zip)
+            regex = "^\[.*\]\((.*)\)$"
+            matches = re.match(regex, attachment["markdown"], flags=re.M)
+            if matches:
+                link = matches.group(1)
 
-        return attachment["markdown"]
+        return "[" + self.file_description + "](" + link + ")"
```
[the remark this answers is not in the document]
Comment (skirill) [reply bugzilla2gitlab/models.py]: Done

File: bugzilla2gitlab/models.py (hunk 4 above)
Comment (skirill) [reply bugzilla2gitlab/models.py]: Done

File: bugzilla2gitlab/models.py (hunk 1 above)
Comment (skirill) [reply bugzilla2gitlab/models.py]: General comment: I realized that the change was not well structured and hard to follow so I've made a refactor. Please take a look. There are now 2 additional methods in Comment class and each method is kept short.

W.r.t. whether the original class member is a filename or description, I think it's a description. When you upload an attachment file, say `script.py` and describe it, say `Sample script`, Bugzilla will show `Sample script` in attachment table but when you click there your browser will download `script.py`.

My motivation was to preserve this way in gitlab. New code extracts file name (`script.py` in this example) from Content-disposition header same way as your browser does, then uses it when uploading the file to gitlab. Gitlab returns attachment markdown with file name used twice, as both the link text and the uploaded file url. We adjust it to use the original file description as the link text. This produces same markdown as in Bugzilla, and same file name is kept when downloading attachments. One side effect (in our deployment) is that localized file descriptions no longer break the migration, as long as file names are in ASCII. This might be typical for non-English speaking teams.

File: bugzilla2gitlab/models.py (hunk 2 above)
Comment (skirill) [reply bugzilla2gitlab/models.py]: Please review the refactored version. I agree that not being able to parse `Content-disposition` is an exception so I made it an Exception.

File: bugzilla2gitlab/models.py (hunk 3 above)
Comment (skirill) [reply bugzilla2gitlab/models.py]: Please review the refactored version.

File: bugzilla2gitlab/models.py [hunk 7]
```diff
@@ -262,17 +268,29 @@ def parse_filename(cls, comment):
...
-        f = {"file": (self.filename, result.content)}
+        f = {"file": (filename, result.content)}
 
         attachment = _perform_request(url, "post", headers=self.headers, files=f, json=True,
                                       dry_run=conf.dry_run)
 
-        if conf.dry_run:
-            return "[attachment]({})".format(self.filename)
+        link = filename
+        if attachment and attachment["markdown"]:
```
[the remark this answers is not in the document]
Comment (skirill) [reply bugzilla2gitlab/models.py]: I made special check for `conf.dry_run` in a refactored version, please review. Also, any issue with missing markdown in response is considered an exception in import process. Here I'm not so confident, we might have returned an empty link instead, but logically it looks like an exception (you can import whole issue but cannot upload attachments => it's safer to break import)

File: bugzilla2gitlab/models.py (hunk 5 above)
Comment (skirill) [reply bugzilla2gitlab/models.py]: I think it's fine too, I've tried to explain this above. Sorry for the original code being unclear.

**Comment by xmunoz** [comment]
This looks great, thanks for the changes and adding unicode support! Just clean up the flake8 error so that the build passes and I'll merge this.
